cmd/juju/cloud/add.go

```diff
@@ -20,6 +20,14 @@ import (
 	"github.com/juju/juju/environs"
 )
 
+type CloudQueryer interface {
+	ParseCloudMetadataFile(path string) (map[string]cloud.Cloud, error)
```

Comment:
Remember I used the phrase, "jedi mind trick", in one of our conversations about IoC? Here is a great example!

What we're really writing here is a command, not the parsing logic or anything to do with `clouds.yaml`. It's purely about interfacing with the user and then deferring logic to other components. Taking in an `io.Reader` is too much detail for a command; it should deal in *concepts*, e.g. "go parse this path" or "what do you know about this cloud?" If you defer that kind of logic to another module, you can implement/test that logic in one spot and use it all over the place. It is a very subtle (at least I think so) but important difference.

This is also why in the tests were placing files onto the file system and then reading them back out and checking against strings. Because we are not decoupled from handling cloud files, because we're not just a dumb command, this is the only way we can test.

And although testing is the first thing everyone brings up -- and it is important and what we're discussing -- it's not the main reason to do it this way. The main reason is so that the *how* is decoupled from the *what*. And you can change the *how* without having to also change this code. Rick and I *just* ran into this situation yesterday.